### PR TITLE
Set _n_models in CompoundModel instances

### LINKED ISSUE #4892: CompositeModel does not inherit n_models properly
This illustrates the problem:

``` python
>>> from astropy.modeling import models
>>> m1 = models.Box1D(amplitude=[1, 1], x_0=[5000, 6000], width=[10, 100], n_models=2)
>>> m1._n_models
2
>>> m2 = models.Trapezoid1D(amplitude=[100, 100], x_0=[5000, 6000], width=[10, 100], slope=[1, 2], n_models=2)
>>> m2._n_models
2
>>> m = m1 + m2
>>> m
<CompoundModel9(amplitude_0=[ 1., 1.], x_0_0=[ 5000., 6000.], width_0=[  10., 100.], amplitude_1=[ 100., 100.], x_0_1=[ 5000., 6000.], width_1=[  10., 100.], slope_1=[ 1., 2.])>
>>> m._n_models
1
```

As a result, evaluation on the compound model breaks:

``` python
>>> x = np.arange(3000, 7000)
>>> m1(x, model_set_axis=False)
array([[ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.]])
>>> m2(x, model_set_axis=False)
array([[ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.]])
>>> m(x, model_set_axis=False)
...
.../astropy/modeling/core.py in __call__(self, x, model_set_axis)
    383             args = ('self',) + inputs
    384             new_call = make_function_with_signature(
--> 385                     __call__, args, [('model_set_axis', None)])
    386             update_wrapper(new_call, cls)
    387             cls.__call__ = new_call

.../astropy/modeling/core.pyc in __call__(self, *inputs, **kwargs)
    379                 """Evaluate this model on the supplied inputs."""
    380 
--> 381                 return super(cls, self).__call__(*inputs, **kwargs)
    382 
    383             args = ('self',) + inputs

.../astropy/modeling/core.pyc in __call__(self, *inputs, **kwargs)
    682 
--> 683         inputs, format_info = self.prepare_inputs(*inputs, **kwargs)
    684         parameters = self._param_sets(raw=True)
    685 

.../astropy/modeling/core.pyc in prepare_inputs(self, *inputs, **kwargs)
   1190         if n_models == 1:
   1191             return _prepare_inputs_single_model(self, params, inputs,
-> 1192                                                 **kwargs)
   1193         else:
   1194             return _prepare_inputs_model_set(self, params, inputs, n_models,

.../astropy/modeling/core.pyc in _prepare_inputs_single_model(model, params, inputs, **kwargs)
   2806                     "broadcast with parameter {2!r} of shape "
   2807                     "{3!r}.".format(model.inputs[idx], input_shape,
-> 2808                                     param.name, param.shape))
   2809 
   2810             if len(broadcast) > len(max_broadcast):

ValueError: Model input argument u'x' of shape (4000,) cannot be broadcast with parameter u'amplitude_0' of shape (2,).
```

### PR DESCRIPTION
**Pull request by mirca**
This is an attempt to fix #4892.

From what I've investigated, there was not a way that a CompositeModel would "inherit" the class variable `_n_models` from model sets that constitute a CompositeModel.

Hence, I'm just assign `_n_models` from one of the model sets to the CompositeModel, providing that the model sets have the same `_n_models`.

@pllim @nden Do you see any drawback in doing this? If that sounds OK, I can go ahead and create an example in the docs page.

### CODE REVIEW

File: astropy/modeling/core.py [hunk 1]
```diff
@@ -2015,6 +2015,12 @@ class members that should be added to the generated
             instance._user_inverse = mcls._make_user_inverse(
                     operator, left, right)
 
+            if left._n_models == right._n_models:
+                instance._n_models = left._n_models
+            else:
+                raise ValueError('Model sets must have the same number of '
+                                 'components.')
```
Comment (pllim) [inline astropy/modeling/core.py]: The error message can be more informative. Perhaps listing the conflicting `_n_models` values.

**Comment by pllim** [comment]
p.s. I am playing with the new GitHub features, don't mean to be so strict with that "requested changes" thingy.

**Comment by nden** [comment]
At a first glance it seems this approach may work but I need to play with this a bit more simply because there are too many options and model sets are not well tested.

One possible complication is the unfortunate use of `len` in two different contexts.
`len(single_model_set)` is the recommended way of getting the value of `model._n_models`.
`len(composite_model)` returns the number of submodels in a composite model. One way around this is to reconstitute the old `model.n_models` attribute for model sets.

Another potential issue is handling inverse models of submodels which are model sets. There may be additional checks needed to handle it.

**Comment by mirca** [comment]
@nden thoughts on this?

**Comment by nden** [comment]
@mirca :+1: and sorry it took so long to get back to this, I swamped with other work.

Not very related to this, while testing this PR I found it impossible to fit model sets any more and I don't see a test for this any more. I wonder if I am doing anything wrong. Have you had a chance to try fitting linear model sets?

**Comment by nden** [comment]
Given that this is a bug is 1.2.2 the correct milestone?

**Comment by pllim** [comment]
If it is merged today, yes; If after the freeze, no?

**Comment by mirca** [comment]
@nden When I tried to fit a linear compound model using `LinearLSQFitter`, #5467 happens. And it seems that `LevMarLSQFitter` does not support fitting model sets.

**Comment by nden** [comment]
I think we can merge this and work on fitting of composite sets and #5467  later.

@pllim Why does the freeze matter? It's still a bug in 1.2.2. Just trying to understand ...

**Comment by mirca** [comment]
@nden @pllim Does this need a changelog entry?

**Comment by nden** [comment]
@mirca Yes, please add a changelog entry under 1.2.2.

**Comment by nden** [comment]
And add [skip ci] in the commit message.

**Comment by pllim** [comment]
@nden , yeah, you are right. Maybe I misunderstood @astrofrog 's reply. Not sure if 1.2.2 is frozen with 1.3. But we can always retag to 1.2.3 or something if needed.

**Comment by mirca** [comment]
@nden Seems that [skip ci] does not affect CircleCI... :(

**Comment by nden** [comment]
yes, it looks that way. The failure is unrelated but I can't restart it. @pllim ?

@pllim I'm also confused about the milestones. If I understand the instructions correctly this should be milestoned as 1.2.2. However, if the changelog is in 1.2.2 will it get into 1.3 if 1.2.2 is not released before 1.3?

**Comment by pllim** [comment]
I don't have CircleCI account. As for milestone, I am not sure either. Maybe @astrofrog can restart the test and answer...

**Comment by bsipocz** [comment]
@nden @pllim - As I understand the plans, there won't be a new 1.2.x release (unless a quick short one for the leap second containing only that change). So at the end of the day everything that was a bugfix, and ended up in the milestone 1.2.2 will be released with 1.3.0. Looking at the 1.2.x branch, actually nothing got backported there yet.
(The LTS 1.0.x branch is different, we are still supposed to support it until 2.0, so probably there will be more 1.0.x releases).

**Comment by nden** [comment]
Thanks for the clarification @bsipocz .

@mirca Could you move the changelog entry to 1.0.11 since it appears this code existed then. Sorry for misjudging the milestone.

**Comment by nden** [comment]
I am going to merge this. circleci failed with a numpy build error and did not honor [skip ci], however previous testing passed.

**Pull request merged**